drivers/modem/modem_at_shell.c

```diff
@@ -0,0 +1,269 @@
+{
+	/* Match anything except the expected response without progressing script */
+	modem_chat_match_init(&at_shell_script_chat_matches[0]);
+	modem_chat_match_set_match(&at_shell_script_chat_matches[0], "");
+	modem_chat_match_set_separators(&at_shell_script_chat_matches[0], "");
+	modem_chat_match_set_callback(&at_shell_script_chat_matches[0], at_shell_print_any_match);
+	modem_chat_match_set_partial(&at_shell_script_chat_matches[0], true);
+	modem_chat_match_enable_wildcards(&at_shell_script_chat_matches[0], false);
+
+	/* Match the expected response and terminate script */
+	modem_chat_match_init(&at_shell_script_chat_matches[1]);
+	modem_chat_match_set_match(&at_shell_script_chat_matches[1], "");
+	modem_chat_match_set_separators(&at_shell_script_chat_matches[1], "");
+	modem_chat_match_set_callback(&at_shell_script_chat_matches[1], at_shell_print_match);
+	modem_chat_match_set_partial(&at_shell_script_chat_matches[1], false);
+	modem_chat_match_enable_wildcards(&at_shell_script_chat_matches[1], false);
+
+	modem_chat_script_chat_init(at_shell_script_chat);
+	modem_chat_script_chat_set_response_matches(at_shell_script_chat,
+						    at_shell_script_chat_matches,
+						    ARRAY_SIZE(at_shell_script_chat_matches));
+	modem_chat_script_chat_set_timeout(at_shell_script_chat,
+					   CONFIG_MODEM_AT_SHELL_RESPONSE_TIMEOUT_S);
+}
+
+static void at_shell_init_pipelink(void)
+{
+	modem_pipelink_attach(at_shell_pipelink, at_shell_pipelink_callback, NULL);
+}
+
+static int at_shell_init(void)
+{
+	at_shell_init_work();
+	at_shell_init_chat();
+	at_shell_init_script_chat();
+	at_shell_init_pipelink();
+	return 0;
+}
+
+SYS_INIT(at_shell_init, POST_KERNEL, 99);
+
+static int at_shell_cmd_handler(const struct shell *sh, size_t argc, char **argv)
+{
+	int ret;
+
+	if (argc < 2) {
+		return -EINVAL;
+	}
+
+	if (!atomic_test_bit(&at_shell_state, AT_SHELL_STATE_ATTACHED_BIT)) {
+		shell_error(sh, "modem is not ready");
+		return -EPERM;
+	}
+
+	if (atomic_test_and_set_bit(&at_shell_state, AT_SHELL_STATE_SCRIPT_RUNNING_BIT)) {
+		shell_error(sh, "script is already running");
+		return -EBUSY;
+	}
+
+	strncpy(at_shell_request_buf, argv[1], sizeof(at_shell_request_buf) - 1);
+	ret = modem_chat_script_chat_set_request(at_shell_script_chat, at_shell_request_buf);
+	if (ret < 0) {
+		return -EINVAL;
+	}
+
+	if (argc == 3) {
+		strncpy(at_shell_match_buf, argv[2], sizeof(at_shell_match_buf) - 1);
+	} else {
+		strncpy(at_shell_match_buf, "OK", sizeof(at_shell_match_buf) - 1);
+	}
+
+	ret = modem_chat_match_set_match(&at_shell_script_chat_matches[1], at_shell_match_buf);
+	if (ret < 0) {
+		return -EINVAL;
+	}
+
+	at_shell_active_shell = sh;
+
+	ret = modem_chat_run_script_async(&at_shell_chat, &at_shell_script);
+	if (ret < 0) {
+		shell_error(sh, "failed to start script");
+		atomic_clear_bit(&at_shell_state, AT_SHELL_STATE_SCRIPT_RUNNING_BIT);
+	}
+
+	return ret;
+}
+
+SHELL_CMD_ARG_REGISTER(at, NULL, "at <command> <response>", at_shell_cmd_handler, 1, 2);
```

Comment:
As this is related to modem, should we have it under modem shell, meaning `modem at ....`?

Reply:
Updated to common `modem` shell with `at` as subcommand :) The initial reason for not doing this was to avoid conflicts with the existing `modem` shell which works with some older modem drivers, but as they are mutually exclusive, I called the new modem shell `MODEM_AT_SHELL` for now :)